### PR TITLE
GlobalISel needs fdiv 1 / sqrt(x) to rsq combine

### LINKED ISSUE #64743: GlobalISel needs fdiv 1 / sqrt(x) to rsq combine
Need port of equivalent in https://reviews.llvm.org/D158099

### PR DESCRIPTION
**Pull request by nickleus27**
Fixes #64743

@arsenm @Pierre-vh Could you guys review and let me know if I am headed in the right direction.
1. is the MIR I am trying to match against 
    `%sqrt:_(s16) = contract G_FSQRT %x`
   ` %one:_(s16) = G_FCONSTANT half 1.0`
    `%rsq:_(s16) = contract G_FDIV %one, %sqrt` 
    ?
2. Will the matcher in `AMDGPUCombine.td` match the above MIR and call the function I made called `matchFDivSqrt`?
3. Any advice on what needs to be done in `AMDGPUPostLegalizerCombinerImpl::matchFDivSqrt` would be appreciated. For example, what is the state of MI that is passed in? Is it a single instruction or is it a chain/tree of instructions?

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPUCombine.td
Comment (tschuett) [inline llvm/lib/Target/AMDGPU/AMDGPUCombine.td:38]: You can remove the `G_FSQRT`.

File: llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp
Comment (tschuett) [inline llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp:343]: assert(MI.getOpode() == TargetOpcode::G_FDIV && "expected fdiv");

**Review by Pierre-vh — CHANGES_REQUESTED** [review]
You're indeed trying to match something like this:
```
%sqrt:_(s16) = contract G_FSQRT %x
%one:_(s16) = G_FCONSTANT half 1.0
%rsq:_(s16) = contract G_FDIV %one, %sqrt
```

@arsenm posted a few examples in the ticket. Please add them all to a `.mir` file in `test/CodeGen/AMDGPU/GlobalISel` and use `update_mir_test_checks` to generate check lines. This allows you to test your changes because you can just rebuild llc, regen the checks and see what your new combine does.

When adding things to the combiner, I would recommend doing it in a test-driven way. It's very helpful IMO because you can verify changes fast: implement something, build llc, regen the lines

Lastly, once this all looks good, don't forget to run `check-llvm-codegen` to catch issues :)

File: llvm/lib/Target/AMDGPU/AMDGPUCombine.td
Comment (nickleus27) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:39]: Thank you for your review. I am using this example, and I am making some progress now. However, one error I am getting that I do not understand is `('G_FCONSTANT') is unreachable from the pattern root!`. Any ideas why I would be getting this error? I am not seeing why this pattern would not be reachable.

**Comment by nickleus27** [comment]
@Pierre-VH @tschuett thank you for your reviews. Your comments helped a lot. First, my apologies for force pushing an amend. I know according to the docs I should've pushed a fix up commit due to your reviews. I accidentally amended, but next time I will remember to do fix-up.

My current commit reflects how I see the patch should look according to Pierre’s review, however, the current commit is not working. I have some questions about how TableGen works that I could not figure out from the docs. I get an error that says (‘G_FCONSTANT’) is unreachable from the pattern root! The docs state TableGen uses the DAG datatype. I am wondering if this is the same DAG that is produced from `DAG.viewGraph()` when debugging? I have included a DAG produced for MIR I am trying to match from the aforementioned graph. The image makes me believe that FSQRT would be the root of the pattern, then links with FDIV, and finally with ConstantFP. (In the commit I have FDIV as the root which would be true if looking at the instructions sequentially going bottom up.) How do I achieve this in the match pattern? I have tried different things with no luck. The docs say, “ By using another DAG node as an argument, an arbitrary graph of DAG nodes can be built.” I have tried this, and also had no luck.
graph11824

**Comment by arsenm** [comment]
> > 1. is the MIR I am trying to match against
> >    `%sqrt:_(s16) = contract G_FSQRT %x`
> >    ` %one:_(s16) = G_FCONSTANT half 1.0`
> >    `%rsq:_(s16) = contract G_FDIV %one, %sqrt`
> >    ?
> 
> Slightly off-topic, but why do we need to match against 1.0? Wouldn't it be better to combine `y/sqrt(x)` into `y*rsq(x)` for all `y`, and then later optimize away the multiply if `y` happened to be 1.0?

We're missing folds for that, at least for f64 last I looked. It was on my todo list for math stuff but I never got around to it

File: llvm/lib/Target/AMDGPU/AMDGPUCombine.td
Comment (Pierre-vh) [inline llvm/lib/Target/AMDGPU/AMDGPUCombine.td:42]: indent
```suggestion
         [{ return ${fpimm}.getFPImm()->isExactlyValue(1.0)
               || ${fpimm}.getFPImm()->isExactlyValue(-1.0); }]),
```

File: llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp
Comment (Pierre-vh) [inline llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp:341]: This should already be set so you can delete it - the combiner will automatically set the insertion point to the MI being looked at

File: llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp
Comment (Pierre-vh) [inline llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp:345]:
```suggestion
  B.buildIntrinsic(Intrinsic::amdgcn_rsq, {Dst})
```
This should work I think, maybe you don't even need the `{}` - I don't remember if implicit ctor works in this case

File: llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp
Comment (Pierre-vh) [inline llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp:347]: I'm not sure at all if that's needed (or if it's even correct - could we contract two operations again after applying contract already?)
cc @arsenm / @jayfoad

File: llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir
Comment (Pierre-vh) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir:3]: add some more types (f32/f64) - you can just copy-paste the f16/neg_f16 test for those types

**Review by Pierre-vh — COMMENTED** [review]
Looking very good, thanks - just a few nits left to sort out :smile:

File: llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp]: Seems reasonable to me to copy the flags. Contracting again should be allowed (but I don't know if it would ever happen in practice for rsq). Other flags like nnan and ninf could perhaps be useful too. If you wanted to get really clever you should union (or intersect?) the flags from the two operations being contracted.

File: llvm/lib/Target/AMDGPU/AMDGPUCombine.td
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AMDGPUCombine.td:42]: How it is correct to match -1.0 here?

Comment (nickleus27) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:42]: I am a novice, so I am asking in hopes of learning, but what would make this incorrect? This seems to work as far as matching MIR that contains -1/sqrt(x) pattern.

File: llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp
Comment (nickleus27) [reply llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp]: I have removed adding the flags since the initial review. Let me know what the final decision is so I can implement. Since I am trying to learn I would like to ask, what would be the potential harm of adding the flags?

File: llvm/lib/Target/AMDGPU/AMDGPUCombine.td
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:42]: Converting 1/sqrt(x) to rsq(x) seems fine since that is the definition of rsq. Converting -1/sqrt(x) to rsq(x) seems wrong because that's just not what rsq does.

Comment (nickleus27) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:42]: Right, that makes sense. Could we do -1/sqrt(x) to fneg(rsq(x))?

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:42]: Yes, we do try to fold that in other places. AMDGPUCodeGenPrepare handles most of these folds, it just leaves the fastest math paths for codegen.

Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:42]: > Could we do -1/sqrt(x) to fneg(rsq(x))?

Yes, especially since fneg is cheap since it can be folded into lots of other (AMDGPU) instructions. But it does make me wonder again, if you're going to generalise, why not generalise to any value? I.e. y/sqrt(x) -> y*rsq(x)

File: llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp
Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/AMDGPUPostLegalizerCombiner.cpp]: Copying the flag sounds good then :)

Flags are optimization hints to enable assumptions, and if those assumptions are broken - or the flag is added when an assumption isn't verified - you can end up with very weird results https://llvm.org/docs/LangRef.html#fastmath

e.g.: When you have `contract`, you're telling the compiler "in this case, you can assume the result of a fused operation will be the same as the result of two distinct operation". This matters because a distinct multiply + add may round numbers twice - once after each operation, but a fused multiply-add (fma) may only round once.  If the compiler knows the inputs of the operations don't care, then it can create a fma, if it doesn't know that, it needs to emit two instructions instead to ensure the result is correct

File: llvm/lib/Target/AMDGPU/AMDGPUCombine.td
Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:42]: > > Could we do -1/sqrt(x) to fneg(rsq(x))?
> 
> Yes, especially since fneg is cheap since it can be folded into lots of other (AMDGPU) instructions. But it does make me wonder again, if you're going to generalise, why not generalise to any value? I.e. y/sqrt(x) -> y*rsq(x)

I agree generalizing makes sense if you're going to handle more than 1/sqrt - you can simplify the apply function at least to just emit `y * rsq(x)` and let other combine eliminate a `1*rsq`, or transform `-1*rsq` into `fneg(rsq)`

Comment (nickleus27) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:42]: > I agree generalizing makes sense if you're going to handle more than 1/sqrt - you can simplify the apply function at least to just emit `y * rsq(x)` and let other combine eliminate a `1*rsq`, or transform `-1*rsq` into `fneg(rsq)`

So should I re-implement this PR to match `y/sqrt(x)` and have the apply function emit `y * rsq(x)`  and a second combine match `1 * rsq` & `-1 * rsq` to `rsq` & `fneg(rsq)` respectively?

Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:42]: `right_identity_one_fp` will take care of the `1 * rsq`, it seems like we don't have one for `-1.0` to `G_FNEG` though so you can do it in a follow up patch if you want. `X * 1` and `X * -1` are generic simplifications that work for anything in X

So no, you don't need to reimplement much, just make the matcher more generic (don't check for +-1.0) and in the apply, you always emit `rsq(x) * y`. Prefer to emit the `y` on the rhs, that way if it's a constant it's already the canonical form for mul (constants are always on the RHS for commutative ops IIRC)

Comment (nickleus27) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td]: Yeah I would like to add fold for `-1.0` to `G_FNEG` in a follow up patch.

File: llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir
Comment (Pierre-vh) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir:25]: Almost done, just needs a bit more testing

It needs some tests for 1/sqrt and -1/sqrt (ideally for every type too) so we can see what happens in those case, and it might be nice to also have a bit more variance in the numerators - try some values like +-0.5, or bigger numbers like 10?
I'm wondering if 0.5 gets converted to rsq/2

You can just change all these tests to use + or -1, and then add the few extra tests with different numerators at the bottom, no need to have those for every type I think

Comment (nickleus27) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir]: I added in the suggested tests. I can add more if it is better to be on the safe side?

**Review by Pierre-vh — APPROVED** [review]
Looks good to me, but wait for @arsenm to confirm before landing as he created the issue

File: llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir
Comment (Pierre-vh) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir]: It's enough tests, IMO

**Comment by nickleus27** [comment]
@Pierre-vh I am waiting for @arsenm's approval since he opened the issue. But, do you know why I do not have the ability to merge the PR? I do not see the "Merge Pull Request" button. And, do you know if there is something I can do to make the button available to me? That way I do not have to ask others to merge for me? I attached a screen shot of where I would expect to see the button.
Screenshot 2024-02-03 at 2 23 07 PM

File: llvm/lib/Target/AMDGPU/AMDGPUCombine.td
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUCombine.td:39]: Needs more flags. Without the denormal correction code, this isn't more precise. Also switches to dropping denormals and 1ulp. You can refer to the dag version for the exact conditions

**Comment by Pierre-vh** [comment]
> @Pierre-vh I am waiting for @arsenm's approval since he opened the issue. But, do you know why I do not have the ability to merge the PR (because I don't have write access)? I do not see the "Merge Pull Request" button. And, do you know if there is something I can do to make the button available to me? That way I do not have to ask others to merge for me? I attached a screen shot of where I would expect to see the button. Screenshot 2024-02-03 at 2 23 07 PM

I assume you don't have commit access? Then you won't be able to merge yourself, but we can do it.
Once you have merged a few patches, you can ask for commit access: https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access

File: llvm/lib/Target/AMDGPU/AMDGPUCombine.td
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:39]: Also missing hasOneUse check

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:39]: Actually contract is adequate for the f16 case, f32 is the complicated one

Comment (nickleus27) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:39]: So I should make a match function to check for conditions for dropping denormals and 1ulp? As for the flags, should I also check for the flags in the same cpp function or add them to the GICombineRule?

Comment (nickleus27) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:39]: I am not seeing why / where we need hasOneUse check?

Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:39]: Need to check that `G_FSQRT` has one user, so we don't do the combine if it means the `G_FSQRT` stays.
(also means you will need a test with a `G_FSQRT` that has multiple users)

To check `hasOneUse`, you can use a C++ predicate on top of the pattern that calls `MRI.hasOneUse` on `${sqrt}`

Comment (nickleus27) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td]: oddly enough I added the test, and it seemed to work for whatever reason?

Comment (nickleus27) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td]: Do we not want to do the combine at all if the SQRT has multi-uses?

Comment (nickleus27) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td]: Thanks for the reply.

**Comment by arsenm** [comment]
I think this is close. I forgot that I remove the codegen support for rsq in the DAG over the summer. I think this patch only needs to handle the f16 part, and not introduce f32/f64 rsq. We do that in CodeGenPrepare

File: llvm/lib/Target/AMDGPU/AMDGPUCombine.td
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:38]: I guess we also want the f16 vectors to work too

Comment (nickleus27) [reply llvm/lib/Target/AMDGPU/AMDGPUCombine.td:38]: So don't worry about f16 vectors? just f16:sqrt is okay?

File: llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir:28]: missing tests with the missing contract on each inst

**Review by arsenm — APPROVED** [review]

File: llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir:28]: Probably doesn't really matter, eventually the real testing should be the end to end IR tests shared with the DAG

Comment (nickleus27) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir:28]: I can add in the tests without contract.

Comment (nickleus27) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir:28]: > Probably doesn't really matter, eventually the real testing should be the end to end IR tests shared with the DAG

What is the end to end IR tests shared with the DAG?

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir:28]: e.g. we have llvm/test/CodeGen/AMDGPU/rsq.f32.ll and a bunch of others. I was working on getting consistent coverage here, but not sure I ever got to f16

Comment (nickleus27) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir:28]: Okay, I will work on adding rsq.16.11 and I will use rsq.32.ll as an example.

Comment (nickleus27) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fdiv-sqrt-to-rsq.mir:28]: I see llvm/test/CodeGen/AMDGPU/rfdiv.f16.ll has some rsq tests. Should I add more tests to this file or make a new file rsq.f16.ll?

**Comment by nickleus27** [comment]
@arsenm I pushed tests with missing contract instructions for f16 type in the .mir file. I looked into adding .ll file tests for f16 type rsq. I found f16 rsq tests in llvm/test/CodeGen/AMDGPU/fdiv.f16.ll which included tests for f16 type with and without contract. I am not sure what I would need to do to make the tests applicable to this patch. Any suggestions?

**Comment by nickleus27** [comment]
If everyone approves could someone go ahead and commit for me. I do not have write access. Thank you.

**Comment by nickleus27** [comment]
Since #80526 landed first. I need to squash these fixup commits, and pull from upstream since the aforementioned commit will change the .mir in this patch. So once I update the .mir I will try back for review.

**Comment by nickleus27** [comment]
@arsenm can you land this for me if everything looks good? If not please let me know what needs to be improved.

**Pull request merged**
